pkg/daemon/ceph/client/controller_utils.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2020 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package client to manage a controller utils.
+package client
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/clusterd"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var (
+	// ImmediateRetryResult Return this for a immediate retry of the reconciliation loop with the same request object.
+	ImmediateRetryResult = reconcile.Result{Requeue: true}
+	// WaitForRequeueIfCephClusterNotReadyAfter requeue after 10sec if the operator is not ready
+	WaitForRequeueIfCephClusterNotReadyAfter = 10 * time.Second
+	// WaitForRequeueIfCephClusterNotReady waits for the CephCluster to be ready
+	WaitForRequeueIfCephClusterNotReady = reconcile.Result{Requeue: true, RequeueAfter: WaitForRequeueIfCephClusterNotReadyAfter}
+)
+
+// IsReadyToReconcile determines if a controller is ready to reconcile or not
+func IsReadyToReconcile(client client.Client, clustercontext *clusterd.Context, namespacedName types.NamespacedName) (cephv1.ClusterSpec, bool, reconcile.Result) {
+	namespacedName.Name = namespacedName.Namespace
+
+	// Running ceph commands won't work and the controller will keep re-queuing so I believe it's fine not to check
+	// Make sure a CephCluster exists before doing anything
+	cephCluster := &cephv1.CephCluster{}
+	err := client.Get(context.TODO(), namespacedName, cephCluster)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			logger.Errorf("CephCluster resource %q not found in namespace %q", namespacedName.Name, namespacedName.Namespace)
+			return cephv1.ClusterSpec{}, false, ImmediateRetryResult
+		} else if err != nil {
+			logger.Errorf("failed to fetch CephCluster %v", err)
+			return cephv1.ClusterSpec{}, false, ImmediateRetryResult
+		}
+	}
+
+	// If the cluster is ready
+	// Not using k8sutil.ReadyStatus to avoid import cycles
+	if cephCluster.Status.Phase == "Ready" {
```

Comment:
Looks like this method will return true if the CR exists and it is not ready, or if it is ready and the ceph status succeeds. Before this line, seems like we need to return false if the phase is not ready.

Reply:
I believe the Phase will change every time an orchestration is in progress. Let's see if we can use the new conditions to know if the cluster is in a completed state.

Reply:
Yes, if the op restarts the CephCluster state will change because of its reconciliation loop, so we would need some time to get the Ready state, currently not a big deal though.